xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -0,0 +1,471 @@
+    private final OrcaPerRequestReportListener perRpcListener = this::onLoadReport;
+    private volatile long lastUpdated;
+    private volatile long nonEmptySince;
+    private volatile double weight;
+    private volatile WeightedRoundRobinLoadBalancerConfig config;
+
+    WrrSubchannel(Subchannel delegate, TimeProvider timeProvider) {
+      this.delegate = checkNotNull(delegate, "delegate");
+      this.timeProvider = checkNotNull(timeProvider, "timeProvider");
+    }
+
+    private void setConfig(WeightedRoundRobinLoadBalancerConfig config) {
+      this.config = config;
+    }
+
+    @VisibleForTesting
+    void onLoadReport(MetricReport report) {
+      double newWeight = report.getCpuUtilization() == 0 ? 0 :
+              report.getQps() / report.getCpuUtilization();
+      if (newWeight == 0) {
+        return;
+      }
+      if (nonEmptySince == Integer.MAX_VALUE) {
+        nonEmptySince = timeProvider.currentTimeNanos();
+      }
+      lastUpdated = timeProvider.currentTimeNanos();
+      weight = newWeight;
+    }
+
+    @Override
+    public void start(SubchannelStateListener listener) {
+      delegate().start(new SubchannelStateListener() {
+        @Override
+        public void onSubchannelState(ConnectivityStateInfo newState) {
+          if (newState.getState().equals(ConnectivityState.READY)) {
+            nonEmptySince = Integer.MAX_VALUE;
+          }
+          listener.onSubchannelState(newState);
+        }
+      });
+    }
+
+    private double getWeight() {
+      if (config == null) {
+        return 0;
+      }
+      double now = timeProvider.currentTimeNanos();
+      if (now - lastUpdated >= config.weightExpirationPeriodNanos) {
+        nonEmptySince = Integer.MAX_VALUE;
+        return 0;
+      } else if (now - nonEmptySince < config.blackoutPeriodNanos
+          && config.blackoutPeriodNanos > 0) {
+        return 0;
+      } else {
+        return weight;
+      }
+    }
+
+    @Override
+    protected Subchannel delegate() {
+      return delegate;
+    }
+  }
+
+  @VisibleForTesting
+  final class WeightedRoundRobinPicker extends ReadyPicker {
+    private final List<Subchannel> list;
+    private final AtomicReference<EdfScheduler> schedulerRef;
+    private volatile boolean rrMode = true;
+
+    WeightedRoundRobinPicker(List<Subchannel> list, int startIndex) {
+      super(checkNotNull(list, "list"), startIndex);
+      Preconditions.checkArgument(!list.isEmpty(), "empty list");
+      this.list = list;
+      this.schedulerRef = new AtomicReference<>(new EdfScheduler(list.size()));
+      updateWeight();
+    }
+
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      if (rrMode) {
+        return super.pickSubchannel(args);
+      }
+      int pickIndex = schedulerRef.get().pick();
+      WrrSubchannel subchannel = (WrrSubchannel) list.get(pickIndex);
+      if (!config.enableOobLoadReport) {
```

Comment:
passing in the constructor - can't config be updated while picker does not change?

Reply:
Ah, _that's_ what I was looking into when I was interrupted. I sent these comments because I had a reply to one of your comments "stuck" behind them, but I had also lost track of where I was.

RR will always create a new Picker, but that picker might not be used. But it _is_ used as long as `isEquivalentTo()` returns `false`.